### PR TITLE
Fix array-bounds warning in DTMuonMillepede

### LINKED ISSUE #43135: [DBG_X] DTMuonMillepede: array subscript ... is outside outside array bounds
In the recent [DBG_X IB](https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_amd64_gcc12/CMSSW_13_3_DBG_X_2023-10-26-2300/Alignment/MuonAlignmentAlgorithms), the compiler emits warnings about accessing array elements outside it's bounds:
```
.../src/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc: In member function 'getCqcMatrix':
  .../src/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:582:44: warning: array subscript 5 is outside array bounds of 'void[40]' [-Warray-bounds]
   582 |             Eta[counterLayer][counterLayer2] = -1.0 / (4.0);
      |                                            ^
.../src/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:574:39: note: at offset 40 into object of size 40 allocated by 'operator new []'
  574 |       Eta[counterLayer] = new double[5];
      |                                       ^
  .../src/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:582:44: warning: array subscript 5 is outside array bounds of 'void[40]' [-Warray-bounds]
   582 |             Eta[counterLayer][counterLayer2] = -1.0 / (4.0);
      |                                            ^
.../src/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:574:39: note: at offset 40 into object of size 40 allocated by 'operator new []'
  574 |       Eta[counterLayer] = new double[5];
      |                                       ^
  .../src/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:582:44: warning: array subscript 6 is outside array bounds of 'void[40]' [-Warray-bounds]
   582 |             Eta[counterLayer][counterLayer2] = -1.0 / (4.0);
      |                                            ^
.../src/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:574:39: note: at offset 48 into object of size 40 allocated by 'operator new []'
  574 |       Eta[counterLayer] = new double[5];
      |                                       ^
  .../src/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:580:44: warning: array subscript 5 is outside array bounds of 'void[40]' [-Warray-bounds]
   580 |             Eta[counterLayer][counterLayer2] = 3.0 / (4.0);
      |                                            ^
.../src/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:574:39: note: at offset 40 into object of size 40 allocated by 'operator new []'
  574 |       Eta[counterLayer] = new double[5];
      |                                       ^
  .../src/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:582:44: warning: array subscript 6 is outside array bounds of 'void[40]' [-Warray-bounds]
   582 |             Eta[counterLayer][counterLayer2] = -1.0 / (4.0);
      |                                            ^
.../src/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:574:39: note: at offset 48 into object of size 40 allocated by 'operator new []'
  574 |       Eta[counterLayer] = new double[5];
      |                                       ^
(... etc ...)
```
(54 warnings total)

I think these warnings are valid: the `Eta` array is defined as `double[12][5]`:
https://github.com/cms-sw/cmssw/blob/a05e5a462c054b985d6bd9e9d85b02e725f3ee1e/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc#L570

https://github.com/cms-sw/cmssw/blob/a05e5a462c054b985d6bd9e9d85b02e725f3ee1e/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc#L574

however the `counterLayer2` index goes from 0 to 11 - https://github.com/cms-sw/cmssw/blob/a05e5a462c054b985d6bd9e9d85b02e725f3ee1e/Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc#L575

So we should either change Eta to be 11x11, or change the `counterLayer2` loop.

### PR DESCRIPTION
**Pull request by iarspider**
#### PR description:

Fix array-bounds warning in DTMuonMillepede by replacing C-style arrays with `std::array`s. Fixes https://github.com/cms-sw/cmssw/issues/43135 . 

#### PR validation:

Bot tests.

### CODE REVIEW

**Comment by iarspider** [comment]
Also resolves two copy-paste errors in the code: [1](https://github.com/cms-sw/cmssw/pull/43313/files#diff-dccccdbe33d6ef8fd9f3e9988b3c38b2e5bb772893000f98a68de3b4bbc43949R577), [2](https://github.com/cms-sw/cmssw/pull/43313/files#diff-dccccdbe33d6ef8fd9f3e9988b3c38b2e5bb772893000f98a68de3b4bbc43949R675) - it makes no sense to check `counterLayer` again (and the rest of the code seems to confirm the assumption of copy-paste error)

**Comment by iarspider** [comment]
please test for CMSSW_14_0_DBG_X

**Comment by iarspider** [comment]
please test for CMSSW_14_0_DBG_X

**Comment by iarspider** [comment]
please test

File: Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc [hunk 1]
```diff
@@ -668,13 +667,12 @@ TMatrixD DTMuonMillepede::getbqcMatrix(int wh, int st, int se) {
         }
       }
     }
-    double **Eta = new double *[12];
-    for (int counterLayer = 0; counterLayer < 12; counterLayer++) {
+    std::array<std::array<double, 12>, 12> Eta{};
+    for (size_t counterLayer = 0; counterLayer < 12; counterLayer++) {
```
Comment (perrotta) [inline Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:671]:
```suggestion
    for (size_t counterLayer = 0; counterLayer < Eta.size(); counterLayer++) {
```

File: Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc [hunk 2]
```diff
@@ -668,13 +667,12 @@ TMatrixD DTMuonMillepede::getbqcMatrix(int wh, int st, int se) {
...
       if (counterLayer > 7 && st == 4)
         continue;
-      Eta[counterLayer] = new double[5];
-      for (int counterLayer2 = 0; counterLayer2 < 12; counterLayer2++) {
-        if (counterLayer > 7 && st == 4)
+      for (size_t counterLayer2 = 0; counterLayer2 < 12; counterLayer2++) {
```
Comment (perrotta) [inline Alignment/MuonAlignmentAlgorithms/src/DTMuonMillepede.cc:674]:
```suggestion
      for (size_t counterLayer2 = 0; counterLayer2 < Eta[counterLayer].size(); counterLayer2++) {
```

**Review by perrotta — COMMENTED** [review]
Taking into account that this code is quite likely unused since ages, I write here this comment aimed to synchronize the code in the two methods where you applied the fixes, but we could even merge this PR without such a synchronization. I will sign it for AlCa, then, and I'll let you @iarspider decide whether to update it or not (if you do so before the time the release managers will manage to merge)

**Comment by perrotta** [comment]
+alca
- See https://github.com/cms-sw/cmssw/pull/43313#pullrequestreview-1737409650

**Comment by iarspider** [comment]
@perrotta thanks for noticing!

**Comment by iarspider** [comment]
please test for CMSSW_14_0_DBG_X

**Comment by perrotta** [comment]
+1

**Comment by rappoccio** [comment]
+1

**Pull request merged**